### ISSUE TITLE
CursorLine overrides QuickFixLine

### ISSUE THREAD

**Issue by romainl** [vim/vim#10654]
### Steps to reproduce

1. Do `$ vim --clean sample`.
2. Add the following content and write the file:

       foo
       bar
       foo
       bar

3. Enable `cursorline`:

       :set cursorline

4. Search for `foo` in the current file:

       :vim foo %

5. Open the quickfix window:

       :cwindow

6. The cursor is on the current item. `CursorLine` overrides `QuickFixLine`:

   <img width="949" alt="Capture d’écran 2022-07-03 à 12 26 11" src="https://user-images.githubusercontent.com/344335/177035519-dad529cb-87fa-4a7f-a4ac-52ef4bd99e98.png">

### Expected behaviour

- `CursorLine` shouldn't override `QuickFixLine`.

  Rationale:

  When `cursorline` is enabled, the current line can be spotted via several visual hints: `CursorLine`, `CursorLineNr`, `CursorLineFold`, `CursorLineSign`, and the cursor. When it is disabled, the current line can still be spotted via the cursor.

  The current quickfix item can only be spotted via one visual hint: `QuickFixLine`. When overridden by `CursorLine`, there is no other visual hint as to which quickfix item is the current one.

  It is not easy to choose between a) keeping `cursorline` while reducing the utility of `QuickFixLine` and b) disabling `cursorline` only to make `QuickFixLine` work correctly.

- With every attribute set to `NONE`, a highlight group should have no visual effect on anything.

  This might be a separate issue but it seems related enough to be mentioned here. With the same setup as above, setting every attribute of `CursorLine` to `NONE` effectively makes it "transparent":

  <img width="946" alt="Capture d’écran 2022-07-03 à 12 30 49" src="https://user-images.githubusercontent.com/344335/177035649-f2c68143-72ca-49dc-b9df-3007d9d4fa90.png">

  but `QuickFixLine` is still overridden.

### Version of Vim

8.2.3455, 8.2.4113

### Environment

- Operating System: macOS 12.4
- Shell: bash
- Terminal emulator: Terminal.app
- `$TERM`: `xterm-256color`

**Comment by brammool** [comment]
Right, the attributes can be combined.  I do think 'cursorline' should overrule QuickFixLine.

**Comment by habamax** [comment]
This implementation actually "breaks" remade colorschemes (and probably many others), making QuickFixLine + CursorLine somewhat less readable:

With default colorscheme:

1. go to vim repo
2. `vim -Nu NONE --cmd 'set bg=dark | set cursorline | hi cursorline ctermbg=236' -c 'grep edit src/edit.c' -c 'copen'`

![image](https://user-images.githubusercontent.com/234774/178557183-54e0ddab-cf36-4dc9-b26f-9b1414487388.png)


desert
![image](https://user-images.githubusercontent.com/234774/178557462-7def0beb-f400-4d26-8f1a-e4a99f77b075.png)

pablo
![image](https://user-images.githubusercontent.com/234774/178557529-b1c58e03-914e-4c5d-8c43-f3fec66b76e4.png)

delek
![image](https://user-images.githubusercontent.com/234774/178557641-670a43c9-a31c-4e81-8dd2-dedcb61c1f07.png)

etc...

**Comment by habamax** [comment]
So I would propose either:

* revert to "Cursorline overrides Quickfixline" or
* make "Quickfixline overrides Cursorline" (do not combine attributes) as @romainl suggested in the OP

**Comment by bfrg** [comment]
I personally use `bold` for `QuickFixLine`, so for me the current behavior works better than before this change 😂
```vim
highlight QuickFixLine cterm=bold gui=bold
```
Screenshot:
![out](https://user-images.githubusercontent.com/6266600/178567454-87c0345a-cf4a-45eb-b807-9a82cd2a18ce.png)

**Comment by habamax** [comment]
Then I guess there is a third variant:

* combine only `gui` and `cterm` attributes.

**Comment by romainl** [comment]
@brammool As of 9.0.54 `CursorLine` and `QuickFixLine` clash even more than previously. It looks like https://github.com/vim/vim/commit/7fe956d17650b231f173868531bc7466010687f0 didn't fix anything.

In my opinion, this issue should be reopened and discussed seriously.

- The initial issue was with `CursorLine` completely overriding `QuickFixLine`. Now that they "combine" it is even worse.

- Combining attributes can only work if the two highlight group have non-overlapping sets of attributes.

  If `CursorLine` only has `cterm=underline`, then it is easy to combine it with the `ctermbg=11` of `Search` because there is no overlap/conflict.

  <img width="987" alt="Capture d’écran 2022-07-16 à 18 17 18" src="https://user-images.githubusercontent.com/344335/179363236-3809bdff-9ece-46be-b21c-c8174e160d79.png">

  But some combinations are bound to clash, as demonstrated upthread.

**Comment by brammool** [comment]
> @brammool As of 9.0.54 `CursorLine` and `QuickFixLine` clash even more than previously. It looks like https://github.com/vim/vim/commit/7fe956d17650b231f173868531bc7466010687f0 didn't fix anything.
> 
> In my opinion, this issue should be reopened and discussed seriously.
> 
> - The initial issue was with `CursorLine` completely overriding `QuickFixLine`. Now that they "combine" it is even worse.
> 
> - Combining attributes can only work if the two highlight group have non-overlapping sets of attributes.
> 
>   If `CursorLine` only has `cterm=underline`, then it is easy to combine it with the `ctermbg=11` of `Search` because there is no overlap/conflict.
> 
>   <img width="987" alt="Capture d’écran 2022-07-16 à 18 17 18" src="https://user-images.githubusercontent.com/344335/179363236-3809bdff-9ece-46be-b21c-c8174e160d79.png">
> 
>   But some combinations are bound to clash, as demonstrated upthread.

OK, we can change this.  So the relevant alternatives are to let
QuickFixLine overrule CursorLine, or perhaps make the combine of the two
reverse.

My worry is that if 'cursorline' is active, then moving the cursor
up/down and crossing the QuickFixLine highlighted line, it's like the
CursorLine highlight disappears.

Combining only part of the highlight isn't really feasible, it likely
creates more problems.

We can try the overruling first and let me know if that works acceptably
well.

-- 
From "know your smileys":
 +<(:-) The Pope

 /// Bram Moolenaar -- ***@***.*** -- http://www.Moolenaar.net   \\\
///                                                                      \\\
\\\        sponsor Vim, vote for features -- http://www.Vim.org/sponsor/ ///
 \\\            help me help AIDS victims -- http://ICCF-Holland.org    ///

**Comment by romainl** [comment]
> My worry is that if 'cursorline' is active, then moving the cursor
up/down and crossing the QuickFixLine highlighted line, it's like the
CursorLine highlight disappears.

As I pointed out in the initial issue, "the curent line" can be spotted with a number of visual hints: `CursorLine`, of course, but also `CursorLineNr`, `CursorLineFold` (less relevant here), `CursorLineSign` (less relevant here), and the cursor itself, of course, so that's a lot of hints.

On the other hand, `QuickFixLine` is the **only** visual hint for "the last quickfix item I jumped to".

- Giving the priority to `CursorLine` effectively kills `QuickFixLine` when they are applied to the same line.
- Giving the priority to `QuickFixLine` effectively removes some utility from `CursorLine` when they are applied to the same line **but it doesn't kill it**.

**Comment by puremourning** [comment]
FWIW I agree with Romain - having QuickFixLine override CursorLine feels reasonably natural. Given the QuickFixLine line is already 'highlighted' it's enough of a visual clue as to where the cursor is I think - if no cursorLine line is visible, then it must be the QuickFixLine line.

**Comment by puremourning** [comment]
I can confirm that reversing the priority on the combined attrs works better IMO:

```diff
ben@BenMBP2021 vim-clean % git diff
diff --git a/src/drawline.c b/src/drawline.c
index eaae4bed3..a5da4c9ad 100644
--- a/src/drawline.c
+++ b/src/drawline.c
@@ -1159,7 +1159,7 @@ win_line(
                else
 # endif
 # if defined(FEAT_QUICKFIX)
-                   line_attr = hl_combine_attr(line_attr, cul_attr);
+                   line_attr = hl_combine_attr(cul_attr, line_attr);
 # else
                    line_attr = cul_attr;
 # endif
```

Before (note the way the quick fix line is unreadable when the cursor is over it): https://asciinema.org/a/PrZAUFPoUmV78ST7ctDU6JkBv

After: https://asciinema.org/a/iwGUr1HHpNacI6i96kHenlOvp

**Comment by romainl** [comment]
IMO, the desirable behavior is:

- when an attribute is present in both CursorLine and QuickFixLine, the QuickFixLine value wins,
- when an attribute is only present in CursorLine, it is applied,
- when an attribute is only present in QuickFixLine, it stays.

**Comment by puremourning** [comment]
@brammool 

> Is that because it only has background highlighting?

Yes I think so.

> If it would have underscore, what is the before/after then?

It's underlined (as the QuickFixLine highlight in question has cterm=NONE): https://asciinema.org/a/aCdFDCi8dngcFzTHIZ7R7DAWW


FWIW, these are the default settings:

```
:hi CursorLine
CursorLine     xxx term=underline ctermbg=236 guibg=#303030

:hi QuickFixLine
QuickFixLine   xxx links to Search
:hi Search
Search         xxx term=reverse ctermfg=235 ctermbg=229 guifg=#262626 guibg=#ffffaf
```

then when I ran `:hi CursorLine cterm=underline`, the above `Search` merges with `CursorLine` to give in effect:

```
term=reverse ctermfg=235 ctermbg=229 guifg=#262626 guibg=#ffffaf cterm=underline
```

I think this effectively applies what @romainl mentioned, but I confess I haven't studied `hl_combine_attr` in a while.

**Comment by habamax** [comment]
Can we have the same for the `cursorcolumn`?

![image](https://user-images.githubusercontent.com/234774/184858802-79390828-002e-4197-b71d-8b9db397747a.png)

![image](https://user-images.githubusercontent.com/234774/184858890-5c49ea32-7189-4ad7-b601-3823d2c3b944.png)

![image](https://user-images.githubusercontent.com/234774/184859056-938d4eea-917a-4a4f-93fb-1bdf9e7195d0.png)

**Comment by brammool** [comment]
> Can we have the same for the `cursorcolumn`?
> 
> ![image](https://user-images.githubusercontent.com/234774/184858802-79390828-002e-4197-b71d-8b9db397747a.png)
> 
> ![image](https://user-images.githubusercontent.com/234774/184858890-5c49ea32-7189-4ad7-b601-3823d2c3b944.png)
> 
> ![image](https://user-images.githubusercontent.com/234774/184859056-938d4eea-917a-4a4f-93fb-1bdf9e7195d0.png)

Unlike 'cursorline' it doesn't have much impact on indicating where the
current entry of the quickfix list is.  Isn't it OK like it is now?

-- 
Q: Is selling software the same as selling hardware?
A: No, good hardware is sold new, good software has already been used by many.

 /// Bram Moolenaar -- ***@***.*** -- http://www.Moolenaar.net   \\\
///                                                                      \\\
\\\        sponsor Vim, vote for features -- http://www.Vim.org/sponsor/ ///
 \\\            help me help AIDS victims -- http://ICCF-Holland.org    ///

**Comment by habamax** [comment]
>  Isn't it OK like it is now?

I guess you're right

**Comment by brammool** [comment]
> > My worry is that if 'cursorline' is active, then moving the cursor
> > up/down and crossing the QuickFixLine highlighted line, it's like the
> > CursorLine highlight disappears.
> 
> As I pointed out in the initial issue, "the curent line" can be
> spotted with a number of visual hints: `CursorLine`, of course, but
> also `CursorLineNr`, `CursorLineFold` (less relevant here),
> `CursorLineSign` (less relevant here), and the cursor itself, of
> course, so that's a lot of hints.

I do not consider that relevant.  When the user has opted to use
'cursorline' the other ways are normally not needed and won't be used.

> On the other hand, `QuickFixLine` is the **only** visual hint for "the
> last quickfix item I jumped to".
> 
> - Giving the priority to `CursorLine` effectively kills `QuickFixLine` when they are applied to the same line.
> - Giving the priority to `QuickFixLine` effectively removes some
> utility from `CursorLine` when they are applied to the same line **but
> it doesn't kill it**.

You could try that by changing the line in src/drawline.c

	line_attr = hl_combine_attr(line_attr, cul_attr);

to:

	line_attr = hl_combine_attr(cul_attr, line_attr);

-- 
There's no place like $(HOME)!

 /// Bram Moolenaar -- ***@***.*** -- http://www.Moolenaar.net   \\\
///                                                                      \\\
\\\        sponsor Vim, vote for features -- http://www.Vim.org/sponsor/ ///
 \\\            help me help AIDS victims -- http://ICCF-Holland.org    ///
